binary_gentoo/internal/cli/tree_diff.py

```diff
@@ -3,12 +3,72 @@
 
 import filecmp
 import os
+import re
 import sys
 from argparse import ArgumentParser
+from typing import Iterable, Set
 
-from ..reporter import exception_reporting
+from ..reporter import announce_and_check_output, exception_reporting
 from ._parser import add_version_argument_to
 
+_keywords_pattern = re.compile('KEYWORDS="(?P<keywords>[^"]*)"')
+
+
+def _get_relevant_keywords_set_for(ebuild_content: str,
+                                   accept_keywords: Iterable[str]) -> Set[str]:
+    match = _keywords_pattern.search(ebuild_content)
+    if match is None:
+        # if the KEYWORDS variable is not found in the ebuild,
+        # we will assume the ebuild needs to be listed
+        ebuild_keywords = accept_keywords
+    else:
+        ebuild_keywords = match.group('keywords')
+        ebuild_keywords = [kw for kw in ebuild_keywords.split(" ") if kw]
+
+    # replace special keywords by the relevant keywords from the ebuild
+    if '**' in accept_keywords:
+        accept_keywords = ebuild_keywords
+    elif '*' in accept_keywords:
+        accept_keywords = [kw for kw in ebuild_keywords if not kw.startswith('~')]
+    elif '~*' in accept_keywords:
+        accept_keywords = [kw for kw in ebuild_keywords if kw.startswith('~')]
+
+    return set(accept_keywords) & set(ebuild_keywords)
+
+
+def _keywords_included_in_ebuild(accept_keywords: Iterable[str],
+                                 new_portdir_ebuild_file: str,
+                                 old_portdir_ebuild_file: str = None) -> bool:
+    """
+    Check if the new_portdir_ebuild_file contains the accept_keywords. Returns True if that is the
+    case, otherwise False. If old_portdir_ebuild_file is passed, an extra comparison is made
+    between the relevant keywords in the old and the new ebuild. If these keywords are identical,
+    return False.
+    """
```

Comment:
I'm leaning towards suggest to make `_keywords_included_in_ebuild` look at one single file only, as the name would apply to some extent, already. The task of this function is rather simple, but the code looks quite complex, so there is probably a chance to get not add as much new complexity.

Reply:
I was thinking about that as well.. Let me have a look at it.

Reply:
I have reorganized a bit:
* reading the ebuild content is now included in the `_get_relevant_keywords_set_for` function
* I simply removed the `_keywords_included_in_ebuild` function and put its content back in `iterate_new_and_changed_ebuilds`. Because reading the ebuild content is moved, that is in fact concise and readable

IMHO this still keeps the function `iterate_new_and_changed_ebuilds` easy to read, and it has the advantage of putting al the "do we output the ebuild or not" logic in one single place. Also, splitting of the `_keywords_included_in_ebuild` function needs at least four parameters (keywords, all-changes/report-changes, old file, new file) so I think it's less readable and more difficult to maintain if it is split out.

Let me know what you think.